### DISCUSSION TITLE
We should follow `laravel/framework` descision that using `@template-covariant` in collections

### DISCUSSION THREAD

**Discussion by mpyw** [larastan/larastan#1697] [General]
### Description

- https://github.com/nunomaduro/larastan/pull/1682#issuecomment-1628965133

@canvural 

I appreciate your diligent work on Larastan and understand your reasons for considering Laravel's decision to make the templates covariant as misguided. However, I would like to put forward a counter-argument for your consideration.

Primarily, Larastan's role, as I understand it, should be to **compensate for Laravel's type definition deficiencies, rather than intentionally override explicitly designed type definitions**. If it goes beyond this, it could lead to unexpected behaviour and confusion for developers who rely on Laravel's type definitions.

Secondly, even TypeScript - a language known for its strong stance on type safety - permits covariance for mutable arrays. This demonstrates that even in mutable contexts, the benefits of covariance can be prioritized. Are you suggesting that this design philosophy, employed by TypeScript, is also fundamentally flawed?

At least, I would like to request the possibility of adding a configuration option that allows users to easily customize the behavior in line with Laravel's approach. In my project, where classes and interfaces are defined separately, I am finding it inconvenient when the `TValue` of a `LazyCollection` returned by a class is not covariant. It forces me to write unnecessary map functions to cast up, such as `$collection->map(fn (ItemInterface $item) => $item)`, which I believe could be avoided if covariance were acknowledged.

Thank you for your time and consideration. I look forward to further discussion on this topic.

**Comment by mpyw** [comment]
CC @Grldk @szepeviktor @mad-briller

**Comment by mpyw** [comment]
I apologize for the slight confusion - I mistakenly opened this as an issue, when it should have ideally been a discussion. I appreciate your understanding.

**Comment by szepeviktor** [comment]
@mpyw Welcome to Discussions!

**Reply by mpyw** [reply]
@szepeviktor Thanks for migrating!

**Comment by mad-briller** [comment]
`@template-covariant` prevents use of the template value in parameter positions, as it breaks liskov substitution principle:

https://phpstan.org/r/d3356a6a-3443-44fb-8036-ea8642d28e90

as such i think laravel's choice to use `@template-covariant` is short sighted, and i also don't believe typescript is a beacon of type system design

you can read more about it in the prs / issues / discussions that are authored / answered by @jiripudil in the phpstan-src repository
i think your issues could be solved by type projections:

https://github.com/phpstan/phpdoc-parser/pull/138#issue-1276575860

**Reply by mpyw** [reply]
@mad-briller I'm well aware of *Liskov's substitution principle*, and I completely understand that from a computer science perspective, what you're saying is correct. However, it's also important to acknowledge that in practical scenarios, such as when working with Laravel's collections, certain responsibilities fall on the programmers.

If mutable operations like push are executed on a Collection, it's the programmer's responsibility to ensure that the code doesn't break. While this might not align with the approach many PHPStan users prefer, I believe many Laravel users would concur. Ideally, the Collection should implement an interface that only allows for immutable operations, and should only be covariant when passed as an immutable collection type. However, given Laravel's current state, I believe the existing approach is sufficient and pragmatic.

Therefore, I believe it becomes problematic when Larastan negates and effectively overrides what Laravel has explicitly written as `@template-covariant`. Many Laravel users, I feel, value this kind of practicality and convenience over strict adherence to computer science principles. This is evident from just observing our community. On the other hand, I can see that PHPStan users prioritize robustness.

As a result, Larastan, bridging these two different perspectives, finds itself in a tricky position, having to listen to the voices of both user groups. While I do understand the viewpoint that says, **"We should strictly adhere to Liskov's substitution principle according to PHPStan's philosophy"**, I think it's also a bit too far to say, **"Laravel made a strange decision so we should nullify it"**.

Regardless of what the default behavior is, I would at least like the option to choose the behavior. Offering this as an option will bring benefits to all users with different philosophies, as it will allow them to switch the behavior according to their project's design.

**Reply by mpyw** [reply]
If Laravel decides to retract the addition of `@template-covariant`, I am willing to step back and accept it. However, what would be most disadvantageous for the community is if the philosophies of Laravel and Larastan were to clash, causing confusion. I reiterate, there should at least be the option to choose. Enforcing invariance without an option is undoubtedly disadvantageous for Larastan users who align more with Laravel's philosophy.

**Comment by mpyw** [comment]
One possible approach could be to define an **`ImmutableCollection` (or `ReadonlyCollection`) interface** that excludes the destructive operations of Laravel's `Collection`. Within the scope of this type, we might be able to consider `TValue` as covariant, without needing to alter the `Collection` itself.

However, I'm not sure if a class implementing a covariant interface can declare an invariant `TValue`. So, there's a possibility that this may not function as expected. Also, it might become inconvenient when trying to process inline without passing through areas that are forced to convert due to the function return or argument type specifications, as you would still need to convert the type of the collection itself. Therefore, it's uncertain whether this could be a complete solution.

If this approach is technically feasible and if it's considered worthwhile to implement, despite the aforementioned imperfections, it might be worth proposing to the [laravel/framework](https://github.com/laravel/framework) side. However, if that proposal were to be rejected, I believe the only other option would be to introduce an option to Larastan. I'm certain that the day when Laravel's `Collection` becomes immutable (or read-only) at the implementation level will never come.

**Comment by KentarouTakeda** [comment]
I agree with @mpyw .

Correctness as a type system is of course important, but it doesn't make sense for Laravel and Larastan to invalidate each other's features.

See for example the [issue for the tool laravel-ide-helper](https://github.com/barryvdh/laravel-ide-helper/issues). This tool isn't as active as Larastan, so **it doesn't exactly follow Laravel's specs, and that seems to confuse many users.**

Laravel and Larastan should complement each other. should not interfere.

**Comment by mpyw** [comment]
As the author of Larastan, I'm also interested in hearing @nunomaduro's thoughts on this matter.

**Comment by canvural** [comment]
Hi there,

Thanks for bringing this up for discussion.

I think we can agree that marking a mutable collection as covariant might not be the best choice. There are quite a few examples and articles that highlight this concern, which I won't repeat here, as you mentioned earlier. Instead, Laravel itself could consider following the approach used by Doctrine collections, or any other package, and extract a read-only interface.

Regarding Larastan and Laravel having some differences, I don't see it as a major issue. Larastan has already fixed many incorrect type hints in Laravel with stub files, and I believe this is a similar situation.

Lastly, it seems like there might not have been too much thought put into accepting that PR in Laravel 😅. If any changes are necessary, I think it would be reasonable for Laravel to consider reverting the PR 🤷🏽‍♂️.

**Comment by canvural** [comment]
Additionally, with the upcoming user side type projections in PHPStan, there might be a solution to this problem.

**Reply by mikethea1** [reply]
@canvural can you explain further?

**Reply by mad-briller** [reply]
@mikethea1 this phpstan article should help;
https://phpstan.org/blog/guide-to-call-site-generic-variance

**Reply by mikethea1** [reply]
@mad-briller thanks I understand now. 

I agree that this helps, but in practice it requires me to mark nearly every method that takes in a collection with `covariant`. The reason is that often phpstan will infer types like `Collection<int, SomeEnum::Value1|SomeEnum::Value2>` and then complain that it can't be passed to a method expecting `Collection<int, SomeEnum>` (many other examples like this e.g. inferring `int<1, max>` instead of `int`).

I ended up using a composer script to patch the larastan stubs to match laravel since that solution felt less verbose even though it is ultimately less safe in edge-cases.
